### PR TITLE
provider/aws: add aws_ssm_association resource

### PR DESCRIPTION
**Pull request by liamjbennett**
Depends on https://github.com/hashicorp/terraform/pull/8282

Part of solution for: https://github.com/hashicorp/terraform/issues/4952

### CODE REVIEW

**Comment by liamjbennett** [comment]
```
➜  terraform git:(aws_ssm_association) ✗ TF_LOG=TRACE TF_LOG_PATH=test_log.txt make testacc TEST=./builtin/providers/aws TESTARGS='-run=TestAccAWSSSMAssociation_basic'
==> Checking that code complies with gofmt requirements...
/home/liam/development/go/bin/stringer
go generate $(go list ./... | grep -v /terraform/vendor/)
2016/08/25 15:02:37 Generated command/internal_plugin_list.go
TF_ACC=1 go test ./builtin/providers/aws -v -run=TestAccAWSSSMAssociation_basic -timeout 120m
=== RUN   TestAccAWSSSMAssociation_basic
--- PASS: TestAccAWSSSMAssociation_basic (195.93s)
PASS
ok      github.com/hashicorp/terraform/builtin/providers/aws    195.941s
```

**Comment by liamjbennett** [comment]
```
➜  terraform git:(aws_ssm_association) TF_LOG=TRACE TF_LOG_PATH=test_log.txt make testacc TEST=./builtin/providers/aws TESTARGS='-run=TestAccAWSSSMAssociation_basic'
==> Checking that code complies with gofmt requirements...
/home/liam/development/go/bin/stringer
go generate $(go list ./... | grep -v /terraform/vendor/)
2016/08/25 20:33:38 Generated command/internal_plugin_list.go
TF_ACC=1 go test ./builtin/providers/aws -v -run=TestAccAWSSSMAssociation_basic -timeout 120m
=== RUN   TestAccAWSSSMAssociation_basic
--- PASS: TestAccAWSSSMAssociation_basic (271.43s)
PASS
ok      github.com/hashicorp/terraform/builtin/providers/aws    271.433s
```

File: builtin/providers/aws/resource_aws_ssm_association.go [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+				Required: true,
+			},
+			"name": {
+				Type:     schema.TypeString,
+				ForceNew: true,
+				Required: true,
+			},
+			"parameters": &schema.Schema{
+				Type:     schema.TypeMap,
+				Optional: true,
```
Comment (stack72) [inline builtin/providers/aws/resource_aws_ssm_association.go]: If this isn't updatable, then it should be marked as ForceNew: true,

File: builtin/providers/aws/resource_aws_ssm_association.go [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+	}
+
+	if v, ok := d.GetOk("parameters"); ok {
+		assosciationInput.Parameters = expandDocumentParameters(v.(map[string]interface{}))
+	}
+
+	resp, err := ssmconn.CreateAssociation(assosciationInput)
+
+	if err != nil {
+		l, _ := ssmconn.ListDocuments(nil)
```
Comment (stack72) [inline builtin/providers/aws/resource_aws_ssm_association.go]: why would we want to make an API call if there was an error?

File: builtin/providers/aws/resource_aws_ssm_association.go [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+
+		log.Printf("[DEBUG] documents found %s", l.DocumentIdentifiers)
+
+		return errwrap.Wrapf("[ERROR] Error creating SSM association: {{err}}", err)
+	}
+
+	d.SetId(*resp.AssociationDescription.Name)
```
Comment (stack72) [inline builtin/providers/aws/resource_aws_ssm_association.go]: Can we guarantee that AssociationDescription isn't nil?

File: builtin/providers/aws/resource_aws_ssm_association.go [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+		InstanceId: aws.String(d.Get("instance_id").(string)),
+	}
+
+	resp, err := ssmconn.DescribeAssociation(params)
+
+	if err != nil {
+		return errwrap.Wrapf("[ERROR] Error reading SSM association: {{err}}", err)
+	}
+
+	association := resp.AssociationDescription
```
Comment (stack72) [inline builtin/providers/aws/resource_aws_ssm_association.go]: can we guarantee that this isn't nil?

File: builtin/providers/aws/resource_aws_ssm_association.go [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+	log.Printf("[DEBUG] Reading SSM Assosciation: %s", d.Id())
+
+	params := &ssm.DescribeAssociationInput{
+		Name:       aws.String(d.Get("name").(string)),
+		InstanceId: aws.String(d.Get("instance_id").(string)),
+	}
+
+	resp, err := ssmconn.DescribeAssociation(params)
+
+	if err != nil {
```
Comment (stack72) [inline builtin/providers/aws/resource_aws_ssm_association.go]: Is there a chance that someone will manually delete the association from the console? What will happen if so?

File: builtin/providers/aws/resource_aws_ssm_association_test.go [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+		conn := testAccProvider.Meta().(*AWSClient).ssmconn
+
+		_, err := conn.DescribeAssociation(&ssm.DescribeAssociationInput{
+			Name:       aws.String(rs.Primary.Attributes["name"]),
+			InstanceId: aws.String(rs.Primary.Attributes["instance_id"]),
+		})
+
+		if err != nil {
+			return err
+		}
```
Comment (stack72) [inline builtin/providers/aws/resource_aws_ssm_association_test.go]: what do we want the test behaviour to be if there is an empty response for DescribeAssociation?

File: builtin/providers/aws/resource_aws_ssm_association_test.go [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+
+func testAccCheckAWSSSMAssociationDestroy(s *terraform.State) error {
+	conn := testAccProvider.Meta().(*AWSClient).ssmconn
+
+	for _, rs := range s.RootModule().Resources {
+		if rs.Type != "aws_ssm_association" {
+			continue
+		}
+
+		out, err := conn.DescribeAssociation(&ssm.DescribeAssociationInput{
```
Comment (stack72) [inline builtin/providers/aws/resource_aws_ssm_association_test.go]: what is out is nil or has a nil Association?

File: builtin/providers/aws/resource_aws_ssm_association_test.go [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+		return nil
+	}
+
+	return fmt.Errorf("Default error in SSM Assosciation Test")
+}
+
+func testAccAWSSSMAssociationBasicConfig(rName string) string {
+	return fmt.Sprintf(`
+resource "aws_security_group" "tf_test_foo" {
+  name = "tf_test_foo"
```
Comment (stack72) [inline builtin/providers/aws/resource_aws_ssm_association_test.go]: we should have %s here too IMO

File: builtin/providers/aws/resource_aws_ssm_association_test.go [hunk 9]
```diff
@@ -0,0 +1,138 @@
...
+    }
+  }
+DOC
+}
+
+resource "aws_ssm_association" "foo" {
+  name        = "test_document_association-%s",
+  instance_id = "${aws_instance.foo.id}"
+}
+`, rName, rName)
```
Comment (stack72) [inline builtin/providers/aws/resource_aws_ssm_association_test.go]: if we add %s to security_group then we need another rName here

File: website/source/docs/providers/aws/r/ssm_association.html.markdown [hunk 10]
```diff
@@ -0,0 +1,80 @@
+---
+layout: "aws"
+page_title: "AWS: aws_ssm_association"
+sidebar_current: "docs-aws-resource-ssm-association"
+description: |-
+  Assosciates an SSM Document to an instance.
+---
+
+# aws\_ssm\_document
```
Comment (stack72) [inline website/source/docs/providers/aws/r/ssm_association.html.markdown]: should have association on it

File: builtin/providers/aws/resource_aws_ssm_association.go [hunk 11]
```diff
@@ -0,0 +1,121 @@
...
+		return errwrap.Wrapf("[ERROR] Error reading SSM association: {{err}}", err)
+	}
+
+	association := resp.AssociationDescription
+
+	d.Set("date", association.Date)
```
Comment (stack72) [inline builtin/providers/aws/resource_aws_ssm_association.go]: Isn't defined in the schema

File: website/source/docs/providers/aws/r/ssm_association.html.markdown [hunk 12]
```diff
@@ -0,0 +1,80 @@
...
+* `parameters` - (Optional) Additional parameters to pass to the SSM document.
+
+## Attributes Reference
+
+The following attributes are exported:
+
+* `name` - The name of the SSM document to apply.
+* `instance_ids` - The instance id that the SSM document was applied to.
+* `parameters` - Additional parameters passed to the SSM document.
+* `date`- The date when the association was made.
```
Comment (stack72) [inline website/source/docs/providers/aws/r/ssm_association.html.markdown]: Isn't defined in the schema

**Comment by liamjbennett** [comment]
```
➜  terraform git:(aws_ssm_association) ✗ AWS_REGION=eu-west-1 TF_LOG=TRACE TF_LOG_PATH=test_log.txt make testacc TEST=./builtin/providers/aws TESTARGS='-run=TestAccAWSSSMAssociation_basic'
==> Checking that code complies with gofmt requirements...
/home/liam/development/go/bin/stringer
go generate $(go list ./... | grep -v /terraform/vendor/)
2016/09/02 08:56:04 Generated command/internal_plugin_list.go
TF_ACC=1 go test ./builtin/providers/aws -v -run=TestAccAWSSSMAssociation_basic -timeout 120m
=== RUN   TestAccAWSSSMAssociation_basic
--- PASS: TestAccAWSSSMAssociation_basic (209.11s)
PASS
ok      github.com/hashicorp/terraform/builtin/providers/aws    209.120s
```

**Comment by liamjbennett** [comment]
@stack72 hopefully this now covers all your comments.

**Comment by stack72** [comment]
Hi @liamjbennett 

this LGTM now :) I made 1 change before merging manually - f9183da2e305aff284eb87523967ebff4622b37f

```
% make testacc TEST=./builtin/providers/aws TESTARGS='-run=TestAccAWSSSMAssociation_'                                                    2 ↵ ✹
==> Checking that code complies with gofmt requirements...
go generate $(go list ./... | grep -v /terraform/vendor/)
2016/09/02 16:56:09 Generated command/internal_plugin_list.go
TF_ACC=1 go test ./builtin/providers/aws -v -run=TestAccAWSSSMAssociation_ -timeout 120m
=== RUN   TestAccAWSSSMAssociation_basic
--- PASS: TestAccAWSSSMAssociation_basic (136.23s)
PASS
ok     	github.com/hashicorp/terraform/builtin/providers/aws   	136.246s
```

**Pull request closed without merging**